common/src/main/java/com/terraformersmc/cinderscapes/mixin/MixinServerWorld.java

```diff
@@ -36,33 +38,38 @@ protected MixinServerWorld(MutableWorldProperties properties, RegistryKey<World>
      */
     @Inject(method="tickIceAndSnow", at = @At(value = "HEAD"), locals = LocalCapture.NO_CAPTURE)
     private void cinderscapes$tickAsh(BlockPos tickPos, CallbackInfo ci) {
+        // Ashy shoals only exists in the nether, why do this iteration on any other dimension
+        if (!getDimensionEntry().matchesKey(DimensionTypes.THE_NETHER)) return;
         if (CinderscapesConfig.INSTANCE.enableAshFall) {
             BlockPos pos = tickPos.mutableCopy();
-            BlockState state = getBlockState(pos);
-            RegistryEntry<Biome> biome = this.getBiome(pos);
 
-            while (pos.getY() < 127 && !(
-                    biome.matchesKey(CinderscapesBiomes.ASHY_SHOALS) &&
-                    state.isSideSolidFullSquare(this, pos, Direction.UP) &&
-                    blockAbove(pos).isIn(CinderscapesBlockTags.ASH_PERMEABLE) &&
-                    this.getBlockState(pos.up()).isAir() &&
-                    CinderscapesBlocks.ASH.getDefaultState().canPlaceAt(this, pos.up()))) {
-                pos = pos.up();
-                state = getBlockState(pos);
-                biome = this.getBiome(pos);
-            }
-
-            if (pos.getY() < 127) {
-                this.setBlockState(pos.up(), CinderscapesBlocks.ASH.getDefaultState());
+            for (; pos.getY() < 127; pos.setY(pos.getY() + 1)) {
+                BlockPos up = pos.up();
+                if (!canPlace(up)) {
+                    continue;
+                } else if (!this.getBiome(up).matchesKey(CinderscapesBiomes.ASHY_SHOALS)) {
+                    continue;
+                }
+                this.setBlockState(up, CinderscapesBlocks.ASH.getDefaultState());
+                break;
             }
         }
     }
 
+    @Unique
+    private boolean canPlace(BlockPos pos){
+        return this.getBlockState(pos).isAir() &&
+                blockAbove(pos).isIn(CinderscapesBlockTags.ASH_PERMEABLE) &&
+                CinderscapesBlocks.ASH.getDefaultState().canPlaceAt(this, pos);
+    }
+
+    @Unique
     private BlockState blockAbove(BlockPos pos) {
-        BlockPos iPos = pos.mutableCopy().up();
+        BlockPos iPos = pos.mutableCopy();
 
+        //up() makes new immutable blockpos per call. This uses the Mutable as a Mutable.
         //noinspection StatementWithEmptyBody
-        for (; isAir(iPos) && iPos.getY() < 127; iPos = iPos.up());
+        for (; isAir(iPos) && iPos.getY() < 127; iPos.setY(iPos.getY() + 1));
```

Comment:
The `BlockPos.Mutable#move` method would better indicate the intent here as well:

```suggestion
        for (; isAir(iPos) && iPos.getY() < 127; iPos.move(Direction.UP));
```

Reply:
same comment as the other one